### PR TITLE
Add /contract endpoints

### PR DESCRIPTION
**Pull request by chris124567**
This adds a `muse` like API to the new siad.  This PR depends on code implemented in https://github.com/SiaFoundation/core/pull/24 so we should wait to merge until that is finalized.

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -46,3 +47,37 @@ type SyncerPeer struct {
 type SyncerConnectRequest struct {
 	NetAddress string `json:"netAddress"`
 }
+
+// A ContractBaseRequest contains the information of the host we are dialing
+// for contract related requests.
+type ContractBaseRequest struct {
+	NetAddress string          `json:"netAddress"`
+	TheirKey   types.PublicKey `json:"theirKey"`
+}
```
Comment (lukechampine) [inline api/api.go:56]: I'd just call this `ScanRequest`, and inline the fields in `FormRequest` and `RenewRequest`

File: api/api.go [hunk 2]
```diff
@@ -46,3 +47,37 @@ type SyncerPeer struct {
...
+type ContractFormResponse struct {
+	Contract renter.Contract     `json:"contract"`
+	Parents  []types.Transaction `json:"parents"`
+}
+
+// A ContractRenewRequest requests that the host renew a contract.
+type ContractRenewRequest struct {
+	ContractBaseRequest
+	RenterKey             types.PrivateKey            `json:"renterKey"`
+	Contract              *types.FileContractRevision `json:"contract"`
```
Comment (lukechampine) [inline api/api.go:79]: why use a pointer here?

File: api/client.go [hunk 3]
```diff
@@ -104,6 +105,24 @@ func (c *Client) SyncerConnect(addr string) (err error) {
 	return
 }
 
+// ContractScan scans a host and retrieves its settings.
+func (c *Client) ContractScan(netAddress string, theirKey types.PublicKey) (resp rhp.HostSettings, err error) {
```
Comment (lukechampine) [inline api/client.go:109]: `theirKey` should be `hostKey` (here and elsewhere)

File: api/client.go [hunk 4]
```diff
@@ -104,6 +105,24 @@ func (c *Client) SyncerConnect(addr string) (err error) {
...
+	err = c.post("/contract/scan", ContractBaseRequest{netAddress, theirKey}, &resp)
+	return
+}
+
+// ContractForm forms a contract.
+func (c *Client) ContractForm(netAddress string, theirKey types.PublicKey, renterKey types.PrivateKey, hostFunds, renterFunds types.Currency, endHeight uint64) (resp ContractFormResponse, err error) {
+	err = c.post("/contract/form", ContractFormRequest{ContractBaseRequest{netAddress, theirKey}, renterKey, hostFunds, renterFunds, endHeight}, &resp)
```
Comment (lukechampine) [inline api/client.go:116]: this is a pretty large struct; best to write it out one-line-per-field

**Comment by chris124567** [comment]
Rebased and added support for host sets.  I just left the import to go.sia.tech/core/renter even though it is no longer present because a similar thing (after it was decided that specific RPC handling code like that shouldn't go in `core`) is being worked on in siad and I will replace those references once it is ported.

**Comment by chris124567** [comment]
Rebased to support the recent renterd PR.  Moved all the API methods to the renterd api module.  The level of code duplication between renterd/siad makes me uncomfortable but I guess there may not be a better way.

File: api/renterd/api.go [hunk 5]
```diff
@@ -49,6 +57,10 @@ type server struct {
 	s  Syncer
 	cm ChainManager
 	tp TransactionPool
+
+	mu        sync.Mutex
+	contracts []rhp.Contract
+	hostSets  map[string][]types.PublicKey
```
Comment (lukechampine) [inline api/renterd/api.go:63]: for now, we can just return the contract in the API response and not save anything server-side. We can skip the host set stuff too; I'm not sure it's the best design

File: api/renterd/api.go [hunk 6]
```diff
@@ -188,6 +200,144 @@ func (s *server) syncerConnectHandler(w http.ResponseWriter, req *http.Request,
...
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	cm := (s.cm).(*chain.Manager)
+	session, err := renter.NewSession(cfr.NetAddress, cfr.HostKey, s.w, s.tp, cm)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```
Comment (lukechampine) [inline api/renterd/api.go:236]: we'll want to add a `defer session.Close()` here

File: api/renterd/api.go [hunk 7]
```diff
@@ -210,5 +360,13 @@ func NewServer(cm ChainManager, s Syncer, w Wallet, tp TransactionPool) http.Han
 	mux.GET("/api/syncer/peers", srv.syncerPeersHandler)
 	mux.POST("/api/syncer/connect", srv.syncerConnectHandler)
 
+	mux.POST("/contract/scan", srv.contractScanHandler)
+	mux.POST("/contract/form", srv.contractFormHandler)
+	mux.POST("/contract/renew", srv.contractRenewHandler)
```
Comment (lukechampine) [inline api/renterd/api.go:365]: needs an `/api` prefix

File: api/renterd/requests.go [hunk 8]
```diff
@@ -44,3 +47,49 @@ type SyncerPeerResponse struct {
...
+	EndHeight   uint64           `json:"endHeight"`
+}
+
+// A ContractFormResponse is the response to /contract/form and
+// /contract/renew.  It contains the formed contract and the parent
+// transactions.
+type ContractFormResponse struct {
+	Contract rhp.Contract        `json:"contract"`
+	Parents  []types.Transaction `json:"parents"`
+}
```
Comment (lukechampine) [inline api/renterd/requests.go:74]: My gut says that renew should have a separate `Response` type, even if they currently contain the same fields.

File: api/renterd/api.go (hunk 6 above)
Comment (chris124567) [reply api/renterd/api.go:222]: As far as I can tell Session does not have a Close method

**Comment by chris124567** [comment]
Tried to resolve the merge conflicts with the Github website and it did this... one second.

**Comment by lukechampine** [comment]
It's actually fine (and preferable) to define `ChainManager` in both places; the real `*chain.Manager` will satisfy both interfaces.

**Comment by chris124567** [comment]
These errors are strange.  I don't get them on my system (go1.16.10 linux/amd64) and it doesn't really make sense to me why 

```
assignment mismatch: 1 variable but s.cm.TipContext returns 2 values
```

is the error for this line (which assigns two variables):

```
    vc, err := s.cm.TipContext()
```
when all definitions of ChainManager are:

```
	// A ChainManager manages blockchain state.
	ChainManager interface {
		TipContext() (consensus.ValidationContext, error)
	}
```

**Comment by chris124567** [comment]
Finally :) Should be good to go.

**Pull request merged**